drivers: dma: add dw axi dma controller driver

pbalsundar opened the pull request:
- add driver for dw axi dma controller
- add sample application to perform memory to memory transfer
- adding driver support for agilex5 platform

pbalsundar committed:
CODEOWNERS: updated codeowner for DMA AXI driver

updated codeowners file

Signed-off-by: Balsundar Ponnusamy <balsundar.ponnusamy@intel.com>

pbalsundar committed:
drivers: dma: add dma driver for designware axi DMA controller

Adding dma driver source for designware axi dma controller

Signed-off-by: Balsundar Ponnusamy <balsundar.ponnusamy@intel.com>

pbalsundar committed:
dts: arm64: intel: add dts node for dma controller for agilex5

add dts support for dma to accomodate dma driver bringup on agilex5

Signed-off-by: Balsundar Ponnusamy <balsundar.ponnusamy@intel.com>

pbalsundar commented:
Hi @teburd, @carlocaione, @tbursztyka, @nashif, @galak, @npitre, @povergoing, @SgrrZhf could you please review and give necessary comments?

carlocaione reviewed (commented):
just a couple on nits, in general LGTM.

teburd reviewed (changes requested):
The registers seem almost identical to the ones in dma_dw_common so I'd like to see why this needs to be a completely new driver.

The use of a work queue and mutex here is a bit different than all the other drivers. DMA is expected to work from interrupt call context and the caller is expected to manage ownership/mutual exclusion requirements.

Can you help me understand why this needs to be a different driver, why a work queue is needed, and why a mutex is being used here?

pbalsundar commented:
Hi @teburd 
Sorry for late response. To answer your question
1. The already existing driver dma_dw_common.c is designed for ahb interface and this PR is for axi interface. The register looks similar in naming convention but the register fields and the register offsets are entirely different. So new driver for dma driver with axi interface
2. Work queue use: We have made an option to the user to choose either to process the callbacks in the interrupt context or the user context through Kconfig configuration. If the dma is used in case of the ethernet cases, user may opt to prcocess the callback in the user context. This is not mandatory that user needs process callback using workqueue, user can configure to process callback in interrupt context.
3. Mutex Use: We are using mutex in the driver so as to avoid any synchronization issues. In the SMP system, multiple independent thread like i3c, i2c, uart, spi or any other IP which uses dma for data transfer can use the dma and its channel. This could bring up the synchronization issues if multiple thread tries to use dma at the same time. To protect the data structures we are using the mutext in the driver

teburd reviewed (commented):
Drivers in the tree today call their callbacks in ISRs, DMA does not provide a syscall interface and isn't expected to be directly used with user mode today. I somewhat doubt it ever could be safely. It's unclear why the work queue option is needed here. A work queue task could always be submitted in the callback if needed.

Drivers in the tree today assume single ownership semantics over channels. This allows calling DMA API functions from an ISR context and avoiding what is likely a programming mistake in trying to share a DMA channel. If there are a few common shared registers that might be contended on those should be protected with a spin lock rather than a mutex/semaphore. Updating a few registers should not be a scheduling point, and shouldn't prevent calls from being possible in the ISR context.

I hope that clarifies my reasoning here.

teburd commented:
From my side I'm working on updating existing drivers and adding documentation to better solidify this behavioral expectation https://github.com/zephyrproject-rtos/zephyr/pull/64982

pbalsundar commented on drivers/dma/Kconfig.dw_axi_dmac:
```diff
@@ -0,0 +1,60 @@
+	help
+	  This flag can be enabled if hardware support Linked List multi-block transfer
+
+config DMA_BOTTOM_HALF_WORK_QUEUE
```
Thought in a different way that the user can always configure the Kconfig option to process the callback in the user context if he has to do the heavy task inside callback. And user can always disable the work queue if user want to process the callback in the isr context. 

But as you mentioned, I agree to the fact that user has many options to design their own logic based on the use cases. I will revert these changes and invoke the callback in the interrupt context

pbalsundar commented on drivers/dma/dma_dw_axi.c:
```diff
@@ -0,0 +1,1002 @@
+	/* user data for dma callback for dma block transfer completion */
+	void *priv_data_blk_tfr;
+	/* mutex for chan data */
+	struct k_mutex ch_mutex_lock;
```
intel_socfpga_agilex5_socdk has 2 instance of dma controller with 4 channels enabled per instance. 
And there can be 48 possible number of users (may not be all active at a time but possible to have more than 8 users at a time depending on use cases)

This is the reason we have implemented mutual exclusion to avoid any synchronization issues

pbalsundar commented on samples/drivers/dma/mem_to_mem/src/main.c:
As in zephyr for each different driver category we do have sample applications and also ztest available under tests directory.

tests directory mainly focuses on the regression tests with different possible combination of test cases(which can include  in case of dma tests: setting dma transfer on different channels, changing the burst len, changing data width, etc).

I think main purpose of sample application would be to give the overview or reference code which will enable the user to know how to write the application

I still would like to keep the sample application here. The sample application is generic and anyone can add support of the board to this sample application as well.

teburd replied:
I think that's a fair enough point, if you'd like to keep the sample please do not make it exclusively for socfpga then. I won't block the PR on a sample that is usable on multiple devices.

pbalsundar commented on drivers/dma/dma_dw_axi.c:
```diff
@@ -0,0 +1,1002 @@
+			chan_data->cfg |= DMA_DW_AXI_CFG_TT_FC(M2P_DMAC);
+			lli_desc->ctl |= DMA_DW_AXI_CTL_SRC_MSIZE(msize_src) |
+					DMA_DW_AXI_CTL_DST_MSIZE(msize_dst);
+			WRITE_BIT(chan_data->cfg, DMA_DW_AXI_CFG_HW_HS_DST_BIT_POS, 0);
```
@carlocaione
We can use sys_clear_bit but, I think sys_* functions are mainly used to write onto the memory mapped registers as internally it will perform a volatile read or write.

In this case, chan_data->cfg is a global variable, so I thought it would be better to use WRITE_BIT

pbalsundar commented:
Hi @teburd, we further had discussion on the synchronization in the driver(with reference to comment in the same [PR](https://github.com/zephyrproject-rtos/zephyr/pull/61444#discussion_r1386760596)).
We finally decide to use the existing implementation after going through the various use cases.

The hooks provided by you in the [PR](https://github.com/zephyrproject-rtos/zephyr/pull/65390) will still useful if the peripheral counts which uses dma is more than the channels supported. We can merge that PR if any dma driver wants to adapt to that in such scenarios.

We are planning to use existing implementation of dma_request_channel to get the channel through the atomic flag marking of the channel and release dma channel using dma_release_channel()

While adding support in the Peripheral driver for dma transfer, we are planning to implement the devmux implementation, which will enable peripheral to use one of the free channel of any dma controller in the board

I will push the required changes with all the comments addressed.

pbalsundar commented:
Hi @teburd, @carlocaione @povergoing, Could you please review the latest requested changes that I pushed.

1. Removed work queue implementation for interrupt handling
2. Removed mutext synchronization

pbalsundar commented:
Hi @carlocaione @teburd @SgrrZhf @nashif could you please review further and give necessary comments?

teburd reviewed (changes requested):
The driver looks really good to me now. My only comment here is related to the added sample and the lack of this board being covered by the existing tests.

Why a new sample, you can run the existing dma tests as samples if that's the goal? Is there something particularly interesting about this mem2mem sample rather than reuse of the loop transfer and scatter gather tests? Please drop the sample which does not add much and instead add this board to those tests, run them, and report back the results here if you can.

pbalsundar committed:
tests: drivers: dma: scatter_gather: adding agilex5 board support

added overlay and conf file for agilex5 board

Signed-off-by: Balsundar Ponnusamy <balsundar.ponnusamy@intel.com>

pbalsundar commented:
I would like to have the stale label removed

decsny commented on dts/bindings/dma/snps,designware-dma-axi.yaml:
```diff
@@ -0,0 +1,36 @@
+  "#dma-cells":
+    const: 1
+
+# #dma-cells : Must be <2>.
```
why does this say dma-cells must be 2 but above dma-cells is const 1?

pbalsundar replied:
Initially design was with two dma-cells. I missed to update the comment properly. I updated now

pbalsundar commented:
Hi @teburd 
I have removed the sample application as this is similar to scatter gather test case. And added support for agilex5 board to the test case

pbalsundar commented:
Hi @carlocaione, @decsny, @povergoing Could you please review and give your inputs?

pbalsundar commented:
@carlocaione, @decsny, @povergoing Could you please review and give your inputs?

pbalsundar commented:
Hi @teburd, could you please help me in adding dev-review label to this PR?

teburd commented:
This week the dev-review meeting is unlikely to happen (US holiday) and we're in the middle of a release. I'm adding the label but you will need to join the dev-review meeting July 11, 2024 to present this PR. Thanks!

pbalsundar commented:
Hi @teburd,
I couldn't attend the meeting last week and I think the dev review label has been removed. what can we do as next step for 2nd review?

MaureenHelm commented:
CI results are quite old, could you please rebase to rerun CI?

gdengi commented on dts/arm64/intel/intel_socfpga_agilex5.dtsi:
```diff
@@ -239,4 +239,39 @@
+		#dma-cells = <1>;
+		reg = <0x10DC0000 0x1000>;
+		interrupt-parent = <&gic>;
+		interrupts = <GIC_SPI 166 IRQ_TYPE_LEVEL
```
we can add these in one single line if they are with in the limit - 100 lines column

pbalsundar commented on drivers/dma/dma_dw_axi.c:
```diff
@@ -0,0 +1,912 @@
+
+LOG_MODULE_REGISTER(dma_designware_axi, CONFIG_DMA_LOG_LEVEL);
+
+#define DEV_CFG(_dev)	((const struct dma_dw_axi_dev_cfg *)(_dev)->config)
```
**Conflict Requirement**

There is one conflict requirement of struct dma_context(from dma.h) and DEVICE_MMIO_RAM(from device_mmio.h) to be a first member of a driver data structure. 

1. 
dma.h expects dma_context structure to be a first member of driver struct data:
```
/**
 * DMA context structure
 * Note: the dma_context shall be the first member
 *       of DMA client driver Data, got by dev->data
 */
struct dma_context {
	/** magic code to identify the context */
	int32_t magic;
	/** number of dma channels */
	int dma_channels;
	/** atomic holding bit flags for each channel to mark as used/unused */
	atomic_t *atomic;
};
```

2. 
DEVICE_MMIO_RAM: 
snippet from device_mmio.h
` * struct foo_driver_data {
 *	DEVICE_MMIO_RAM;       -->expected as first member of driver data structure
 *	int wibble;
 *	...
 * }`

**Solution:**
So Since this is a conflict requirement and these both conditions can't be satisfied at the same time. We have other variation of DEIVICE_MMIO_* called DEVICE_MMIO_NAMED_*.

These DEVICE_MMIO_NAMED* macros can use the struct type of the driver's data structure to access the virtual address stored in struct data. So decided to use DEVICE_MMIO_NAMED_* macros instead of DEVICE_MMIO_* macros

As we have to use the DEVICE_MMIO_NAMED* macros, device_mmio.h internally expects DEV_CFG and DEV_DATA to be defined by driver so that device_mmio.h can access the proper address by typecasting

pbalsundar commented on drivers/dma/dma_dw_axi.c:
```diff
@@ -0,0 +1,912 @@
+	}
+
+	/* enable dma controller and global interrupt bit */
+	sys_write64(DMA_DW_AXI_CFG_INT_EN | DMA_DW_AXI_CFG_EN, reg_base + DMA_DW_AXI_CFGREG);
```
Yes, these are 64-bit wide registers MMIO mapped registers
